File: TeleVentas/Clases/PagoTarjeta.py
from Clases.Pago import Pago 

class PagoTarjeta(Pago):

    def __init__(
        self,
        numero_referencia : int,
        monto : float,
        numero_tarjeta: str,
        titular : str,
        vencimiento : str, 
        cvv : str,

    )-> None:
        super().__init__(numero_referencia, monto)
        self.__numero_tarjeta =numero_tarjeta
        self.__titular = titular
        self.__vencimiento = vencimiento
        self.__cvv = cvv

    
    @property
    def numero_tarjeta(self)-> str:
        return f"**** **** **** {self.__numero_tarjeta[-4:]} "
    
    @property
    def titular(self)-> str:
        return self.__titular
    
    @property
    def vencimiento(self)->str:
        return self.__vencimiento
    
    "metodos"
    
    def validar_tarjeta(self)->bool:
        numero_valido = len(self.__numero_tarjeta) == 16 and self.__numero_tarjeta.isdigit()
        cvv_valido = len(self.__cvv) == 3 and self.__cvv.isdigit()
        vencimiento_valido = len(self.__vencimiento) == 5 and "/" in self.__vencimiento
        titular_valido = bool(self.__titular.strip())

        return numero_valido and cvv_valido and vencimiento_valido and titular_valido
    
    def procesar_pago(self)->bool:
        if not self.validar_tarjeta():
            self.estado ="rechazado"
            print(f"Pago Rechazado, datos de la tarjeta inválidos")
            return False
        
        self.estado = "aprobado"
        print(f"Pago Aprobado, Tarjeta {self.__numero_tarjeta},monto ${self.monto:.2f} ")
        return True
    
    def __str__(self) -> str:
        return (
            f"PagoTarjeta(ref={self.numero_referencia}, "
            f"tarjeta={self.numero_tarjeta}, "
            f"titular={self.__titular}, "
            f"monto={self.monto:.2f}), "
            f"estado={self.estado}"
        )
 
    def __repr__(self) -> str:
        return self.__str__()

File: Zsegunda_Parte_Java/Televentas/Clases/PagoTarjeta.java
package Clases;

/**
 * Representa un pago con tarjeta de credito en TeleVentas.
 *
 * Extiende Pago con los datos especificos de la tarjeta y valida
 * que la informacion sea correcta antes de procesar el cobro.
 */
public class PagoTarjeta extends Pago {

    // Atributos privados
    private String numeroTarjeta;
    private String titular;
    private String vencimiento;
    private String cvv;

    // --- Constructor ---

    /** Inicializa el pago con tarjeta con los datos del titular y la tarjeta. */
    public PagoTarjeta(int numeroReferencia, double monto,
                       String numeroTarjeta, String titular,
                       String vencimiento, String cvv) {
        super(numeroReferencia, monto);
        this.numeroTarjeta = numeroTarjeta;
        this.titular       = titular;
        this.vencimiento   = vencimiento;
        this.cvv           = cvv;
    }

    // --- Getters ---

    /** Retorna el numero de tarjeta enmascarado, mostrando solo los ultimos 4 digitos. */
    public String getNumeroTarjeta() {
        return "**** **** **** " + numeroTarjeta.substring(numeroTarjeta.length() - 4);
    }

    /** Retorna el nombre del titular de la tarjeta. */
    public String getTitular() { return titular; }

    /** Retorna la fecha de vencimiento de la tarjeta. */
    public String getVencimiento() { return vencimiento; }

    // --- Metodos ---

    /** Valida que los datos de la tarjeta tengan el formato correcto. */
    public boolean validarTarjeta() {
        boolean numeroValido     = numeroTarjeta.length() == 16 && numeroTarjeta.matches("\\d+");
        boolean cvvValido        = cvv.length() == 3 && cvv.matches("\\d+");
        boolean vencimientoValido = vencimiento.length() == 5 && vencimiento.contains("/");
        boolean titularValido    = !titular.isBlank();
        return numeroValido && cvvValido && vencimientoValido && titularValido;
    }

    /** Valida la tarjeta y procesa el pago, actualizando el estado segun el resultado. */
    @Override
    public boolean procesarPago() {
        if (!validarTarjeta()) {
            setEstado("rechazado");
            System.out.println("Pago rechazado. Datos de la tarjeta invalidos.");
            return false;
        }
        setEstado("aprobado");
        System.out.printf("Pago aprobado. Tarjeta %s — monto $%,.2f%n",
                getNumeroTarjeta(), getMonto());
        return true;
    }

    @Override
    public String toString() {
        return "  PAGO CON TARJETA\n" +
               "  referencia : " + getNumeroReferencia()               + "\n" +
               "  tarjeta    : " + getNumeroTarjeta()                  + "\n" +
               "  titular    : " + titular                             + "\n" +
               "  vencimiento: " + vencimiento                         + "\n" +
               String.format("  monto      : $%,.2f%n", getMonto())   +
               "  estado     : " + getEstado()                         + "\n";
    }
}
